pkg/controller/flinkapplication/flink_state_machine.go

```diff
@@ -77,23 +78,30 @@ func (s *FlinkStateMachine) updateApplicationPhase(ctx context.Context, applicat
 }
 
 func (s *FlinkStateMachine) shouldRollback(ctx context.Context, application *v1alpha1.FlinkApplication) bool {
-	if application.Status.DeployHash == "" {
+	if application.Status.DeployHash == "" && application.Status.LastSeenError == "" {
 		// TODO: we may want some more sophisticated way of handling this case
 		// there's no previous deploy for this application, so nothing to roll back to
 		return false
 	}
 
-	appLastUpdated := application.Status.LastUpdatedAt
-	if appLastUpdated != nil && !v1alpha1.IsRunningPhase(application.Status.Phase) {
-		elapsedTime := s.clock.Since(appLastUpdated.Time)
-		if s.getStalenessDuration() > 0 && elapsedTime > s.getStalenessDuration() {
-			s.flinkController.LogEvent(ctx, application, "", corev1.EventTypeWarning, fmt.Sprintf("Application failed to progress for %v in the %v phase",
-				elapsedTime, application.Status.Phase))
-
-			return true
-		}
+	if application.Status.LastSeenError == "" {
+		return false
 	}
-	return false
+	// Check if the error is retryable and we have retries left
+	if client.IsErrorRetryable(application.Status.LastSeenError) && application.Status.RetryCount <= maxRetries {
```

Comment:
Is there anyway to implement exponential backoff for retries? Otherwise we may very quickly exhaust our retries without giving the problem time to resolve itself.

Reply:
This has also been addressed, albeit some updates are required to the exponential backoff itself.